### PR TITLE
DM-9086 : implement test namespace and finalize_ hook framework

### PR DESCRIPTION
**Pull request by athornton**
This lays the groundwork for the stuff I am about to do with taking Jonathan out of the technote loop.

One is that we now have a "test" series so I am not polluting the real namespace as we flail around.

Two is that I've broken the plugin-loading functionality out into its own function and added a finalize_ function.  That will be the hook we use to make technote post-commit processing (e.g. registering with keeper) work.

Three, this clearly deserves a version bump.

I want to do this as a separate PR now so that I can roll out this to api.lsst.codes, so we can call it and get new test technotes in lsst-sqre.

### CODE REVIEW

File: uservice_ccutter/plugins/finalize.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
+"""Delegate things-to-do-after-creating-repository to particular project-type
+hooks."""
+from .load_plugin import load_plugin
+
+
+def finalize(templatetype, auth, inputdict):
+    """Dispatch to particular type's finalize_ (note trailing underscore;
+    the theory is that your actual fields won't end with one) method, if
+    it exists.  No-op if it doesn't."""
```
Comment (jonathansick) [inline uservice_ccutter/plugins/finalize.py]: End your docstrings with `"""` on a new line.

File: uservice_ccutter/plugins/load_plugin.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+
+def load_plugin(plugin_name):
+    """Load a named plugin.  Substitute "-" with "_" first.  If the module
+    is already loaded, return it; if not, load it and then return the
+    loaded module."""
+    ctt = "." + plugin_name.lower().replace("-", "_")
+    ctr = ".".join(__name__.split(".")[:-1])
+    modname = ctr + ctt
+    mod = None
```
Comment (jonathansick) [inline uservice_ccutter/plugins/load_plugin.py]: Will `mod` ever be returned as `None`?

Comment (jonathansick) [reply uservice_ccutter/plugins/load_plugin.py]: Could you `return mod` from inside the `try`, thereby obviating needing the `mod = None` line?

File: uservice_ccutter/plugins/load_plugin.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    from importlib import ModuleNotFoundError
+except ImportError:
+    ModuleNotFoundError = ImportError
+
+
+def load_plugin(plugin_name):
+    """Load a named plugin.  Substitute "-" with "_" first.  If the module
+    is already loaded, return it; if not, load it and then return the
+    loaded module."""
+    ctt = "." + plugin_name.lower().replace("-", "_")
```
Comment (jonathansick) [inline uservice_ccutter/plugins/load_plugin.py]: I'd like to see these variable names spelled out.

**Comment by jhoblitt** [comment]
It would be nice to set a meaningful PR title.

File: uservice_ccutter/plugins/substitute.py [hunk 4]
```diff
@@ -15,15 +14,12 @@ def substitute(templatetype, auth, inputdict):
...
     one field that's guaranteed to be present so that we can use its hook to
     drive creation of the github_* fields."""
-    ctt = "." + templatetype.lower().replace("-", "_")
-    ctr = ".".join(__name__.split(".")[:-1])
-    importlib.import_module(ctt, package=ctr)
-    modname = ctr + ctt
-    symbols = [x for x in sys.modules[modname].__dict__ if x[0] != "_"]
+    module = load_plugin(templatetype)
+    symbols = [x for x in module.__dict__ if x[0] != "_"
+               and x != "finalize_"]
```
Comment (jonathansick) [inline uservice_ccutter/plugins/substitute.py]: I foresee some maintainability issues by putting knowledge of `finalize_` here into `substitute`. Maybe it'd be better to filter out any object with a name that also *ends* in an underscore, `x[-1] != '_'`.

Then we could add more hooks like `finalize_` and there wouldn't be any need to update the substitution discovery code.

**Review by jonathansick — CHANGES_REQUESTED** [review]
Overall I like the plugin architecture and implementation of a `finalize_` hook in each project's module namespace. Some suggestions:

- [ ] I think it'd be worth refactoring the modules in `plugins`. There seems to be one class of modules, projects like `lsst_technote_bootstrap.py` and `uservice_bootstrap.py`, and then a second class of modules related to the plugin infrastructure. I think it'd be good to separate those into sub packages.

- [ ] See the comments about filtering `finalize_` in `substitute.py`.

File: uservice_ccutter/plugins/load_plugin.py (hunk 2 above)
Comment (athornton) [reply uservice_ccutter/plugins/load_plugin.py]: I can leave out the mod=None line (Python is not Go), but I would need to add another return if I were to move the return inside the try.

There are two tries there.  If the module is already loaded, return it.  If not, try to load it, and then return it.

Comment (jonathansick) [reply uservice_ccutter/plugins/load_plugin.py]: I'd vote for two returns, but I leave it up to you whether you'd want to change anything here.

File: uservice_ccutter/plugins/substitute.py (hunk 4 above)
Comment (athornton) [reply uservice_ccutter/plugins/substitute.py]: Plugins now has a `projecttypes` subdir where actual project type modules go.

**Review by jonathansick — APPROVED** [review]
Looks good, thanks.

**Pull request merged**